File: 백준/Gold/7576. 토마토/토마토.cc
#include <iostream>
#include <queue>
using namespace std;

int m,n;

int dx[4]={0,0,1,-1};
int dy[4]={1,-1,0,0};
int arr[1001][1001];
int visited[1001][1001]={0};
int maxi=0;
queue<pair<int,int>>q;
int main(){
  bool flag=false;
  cin>>m>>n;

  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      cin>>arr[i][j];
    }
  }

  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      if(arr[i][j]==0){
        flag=true;
        break;
      }
    }
  }

  if(flag==false){
    cout<<'0';
    return 0;
  }
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      if(arr[i][j]==1 && visited[i][j]==0){
        arr[i][j]=1;
        q.push({i,j});
        visited[i][j]=1;
      }
    }
  }
  while(!q.empty()){
    int front=q.front().first;
    int back=q.front().second;

    q.pop();
    for(int k=0;k<4;k++){
      int nx=front+dx[k];
      int ny=back+dy[k];

      if(nx>=0 && ny>=0 && nx<n && ny<m && visited[nx][ny]==0 && arr[nx][ny]==0){
        q.push({nx,ny});
        arr[nx][ny]=arr[front][back]+1;
        visited[nx][ny]=1;
        }
      }
  }
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      if(arr[i][j]==0){
        cout<<"-1";
        return 0;
      }
      else{
        maxi=max(maxi, arr[i][j]);
      }
    }
  }

  cout<<maxi-1;
  return 0;
}

File: 백준/Gold/7576. 토마토/토마토.java
/******************************************************************************

                            Online Java Compiler.
                Code, Compile, Run and Debug java program online.
Write your code in this editor and press "Run" button to execute it.

*******************************************************************************/
import java.util.*;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
public class Main
{
    static int[][] arr;
    static boolean[][] visited;
    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, 1, 0, -1};
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] docs = br.readLine().split(" ");
		int n = Integer.parseInt(docs[0]);
		int m = Integer.parseInt(docs[1]);
		arr = new int[m][n];
		visited = new boolean[m][n];
		for(int i=0;i<m;i++){
		    String[] tmp = br.readLine().split(" ");
		    for(int j=0;j<tmp.length;j++){
		        arr[i][j] = Integer.parseInt(tmp[j]);
		    }
		}
		boolean flag = true;
		for(int i=0;i<m;i++){
		    for(int j=0;j<n;j++){
		        if(arr[i][j] == 0){
		            flag = false;
		            break;
		        }
		    }
		}
		if(flag){
		    System.out.println(0);
		    return;
		}
		Queue<Pair> q = new LinkedList<>();
		for(int i=0;i<m;i++){
		    for(int j=0;j<n;j++){
		        if(arr[i][j] == 1){
		            q.add(new Pair(i, j));
		            visited[i][j] = true;
		        }
		    }
		}
		
		while(!q.isEmpty()){
		    Pair p = q.poll();
		    int x = p.x;
		    int y = p.y;
		    for(int i=0;i<4;i++){
		        int nx = x + dx[i];
		        int ny = y + dy[i];
		        if(nx >= 0 && ny >= 0 && nx < m && ny < n && !visited[nx][ny] && arr[nx][ny] == 0){
		            q.add(new Pair(nx, ny));
		            visited[nx][ny] = true;
		            arr[nx][ny] = arr[x][y] + 1;
		        }
		    }
		}
		int maxi = 0;
		for(int i=0;i<m;i++){
		    for(int j=0;j<n;j++){
		        maxi = Math.max(maxi, arr[i][j]);
		        if(arr[i][j] == 0) {
		            System.out.println(-1);
		            return;
		        }
		    }
		}
		
		System.out.println(maxi - 1);
		
		
	}
	
	public static class Pair{
	    int x;
	    int y;
	    Pair(int x, int y){
	        this.x = x;
	        this.y = y;
	    }
	}
}